rfc/0009-audit-checks.md

```diff
@@ -0,0 +1,218 @@
+  id0(Policy Server start) --> id1{Is in audit mode?}
+  id1 -->|Yes|id7{Are there policies with audit enabled?}
+  id7 -->|Yes|id2(Set timer to trigger next audit check)
+  id2 -->id4(Wait audit check timer run off)
+  id4 -->id5(Run audit checks)
+  id5 -->id6(Report resources violating policies)
+  id6 -->id2
+  id1 -->|No|id3B(Evaluate admission requests)
+  id7 -->|No|id2b(Nothing to do)
+```
+
+An audit checks consist of the following steps:
+1. Get information about which policies have audit checks enabled
+2. Get the resources configured in the policies with audit checks enabled
+3. Send the resources that a policy is configured to validate to it and get the validation results.
+4. Aggregate all the policies validation results
+5. Publish a [PolicyReport](https://github.com/kubernetes-sigs/wg-policy-prototypes/tree/master/policy-report) with the aggregated results
+
+In step 2, after fetching the resources for the first time,  they will be stored in a
+cache and used again in futures audit checks until the *time to live* period expired.
+Triggering another request to the Kubernetes API.
+
+It necessary a configuration to define the *time to live* period used by the cache.
+
+## ClusterAdmissionPolicy and AdmissionPolicy CRDs
+
+It's necessary a change in the policy CRDs to mark a policy to be run in the audit checks.
+The field would be called `audit` and its default value will be `true`.
+Therefore, all the policies will run in the audit checks by default, unless the user
+configure the other way. Example of policies CRDs with the `audit` field:
+
+```yaml
+apiVersion: policies.kubewarden.io/v1alpha2
+kind: ClusterAdmissionPolicy
+metadata:
+  name: psp-capabilities
+spec:
+  policyServer: reserved-instance-for-tenant-a
+  module: registry://ghcr.io/kubewarden/policies/psp-capabilities:v0.1.3
+  rules:
+  - apiGroups: [""]
+    apiVersions: ["v1"]
+    resources: ["pods"]
+    operations:
+    - CREATE
+    - UPDATE
+  mutating: true
+  audit: true
+  settings:
+    allowed_capabilities:
+    - CHOWN
+    required_drop_capabilities:
+    - NET_ADMIN
+```
+
+```yaml
+apiVersion: policies.kubewarden.io/v1alpha2
+kind: AdmissionPolicy
+metadata:
+  name: psp-capabilities
+spec:
+  policyServer: reserved-instance-for-tenant-a
+  module: registry://ghcr.io/kubewarden/policies/psp-capabilities:v0.1.3
+  rules:
+  - apiGroups: [""]
+    apiVersions: ["v1"]
+    resources: ["pods"]
+    operations:
+    - CREATE
+    - UPDATE
+  mutating: true
+  audit: true
+  settings:
+    allowed_capabilities:
+    - CHOWN
+    required_drop_capabilities:
+    - NET_ADMIN
+```
+
+## Policy server and policies communication
+
+To be able to call policies to evaluate the resources already in the cluster,
+it's necessary to add a new fields in the `ValidationRequest` (e.g. `resources`)
+and `ValidationRespose` (e.g. `resources_validations`). In the `ValidationRequest`
+the field is needed to store  a list of the resources which the policy must validate.
```

Comment:
I think It's doable. But if we want to use some kind of mechanism similar to what we have for context aware policies we need to add every resource type in the allowed resources and maintain it. Furthermore, we need to add a cache in the middle. Because I would like to avoid a bad policy hammering the kubernetes API. Does it make sense?